### PR TITLE
Track graph validation state in submission state

### PR DESCRIPTION
**Pull request by jacobwindsor**
dcp-546

- Adds states to SubmissionState for graph validation events:
  - `GRAPH_VALIDATION_REQUESTED`
  - `GRAPH_VALIDATING`
  - `GRAPH_VALID`/`GRAPH_INVALID`
- Renames metadata validation states in SubmissionStates to `METADATA_<state>`
- Adds endpoints for requesting state tracker changes and commiting them
- Removes now unused `graphValidationState`
- Removes now uneccesary switching of `graphValidationState` to `PENDING` when the `submissionState` goes to `DRAFT`
- `graphValidationErrors` on a metadata entity are now removed when the submission state switches to `GRAPH_VALIDATION_REQUESTED`

### CODE REVIEW

File: src/main/java/org/humancellatlas/ingest/core/web/MetadataDocumentResourceProcessor.java [hunk 1]
```diff
@@ -50,9 +49,9 @@
             case DRAFT:
                 return Optional.of(Links.DRAFT_REL);
             case VALIDATING:
-                return Optional.of(Links.VALIDATING_REL);
+                return Optional.of(Links.METADATA_VALIDATING_REL);
```
[the remark this answers is not in the document]
Comment (jacobwindsor) [reply src/main/java/org/humancellatlas/ingest/core/web/MetadataDocumentResourceProcessor.java:52]: I think a switch case is easier to reason about

File: src/main/java/org/humancellatlas/ingest/core/web/Links.java [hunk 2]
```diff
@@ -46,12 +54,20 @@
 
     public static final String COMMIT_DRAFT_REL = "commitDraft";
     public static final String COMMIT_DRAFT_URL = "/commitDraftEvent";
-    public static final String COMMIT_VALIDATING_REL = "commitValidating";
-    public static final String COMMIT_VALIDATING_URL = "/commitValidatingEvent";
-    public static final String COMMIT_VALID_REL = "commitValid";
-    public static final String COMMIT_VALID_URL = "/commitValidEvent";
-    public static final String COMMIT_INVALID_REL = "commitInvalid";
-    public static final String COMMIT_INVALID_URL = "/commitInvalidEvent";
+    public static final String COMMIT_METADATA_VALIDATING_REL = "commitValidating";
```
[the remark this answers is not in the document]
Comment (jacobwindsor) [reply src/main/java/org/humancellatlas/ingest/core/web/Links.java:57]: These are used to build the JSON response, they should be strings I think

File: src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelope.java [hunk 3]
```diff
@@ -61,21 +58,38 @@ private static Logger getLog() {
                 allowedStates.add(SubmissionState.DRAFT);
                 break;
             case DRAFT:
-                allowedStates.add(SubmissionState.VALIDATING);
+                allowedStates.add(SubmissionState.METADATA_VALIDATING);
```
[the remark this answers is not in the document]
Comment (jacobwindsor) [reply src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelope.java:61]: I think switch case is easier to reason about

File: src/main/java/org/humancellatlas/ingest/user/UserController.java [hunk 4]
```diff
@@ -71,9 +71,9 @@ public Summary summary() {
...
-        long validatingubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.VALIDATING, user);
-        long validSubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.VALID, user);
-        long invalidSubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.INVALID, user);
+        long validatingubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.METADATA_VALIDATING, user);
+        long validSubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.METADATA_VALID, user);
+        long invalidSubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.METADATA_INVALID, user);
         long submittedSubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.SUBMITTED, user);
         long processingSubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.PROCESSING, user);
         long cleanupSubmissions = submissionEnvelopeRepository.countBySubmissionStateAndUser(SubmissionState.CLEANUP, user);
```
[the remark this answers is not in the document]
Comment (jacobwindsor) [reply src/main/java/org/humancellatlas/ingest/user/UserController.java:79]: More than 3. It's old code that is not from this PR. I removed them

**Review by aaclan-ebi — APPROVED** [review]

File: src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelope.java [hunk 5]
```diff
@@ -105,48 +119,12 @@ private static Logger getLog() {
...
-        return allowedStates;
-    }
-
-    public List<SubmissionGraphValidationState> allowedGraphValidationStateTransitions() {
-        return allowedGraphValidationStateTransitions(getGraphValidationState());
-    }
-
-
     public void enactStateTransition(SubmissionState targetState) {
         if (this.submissionState != targetState) {
```
[the remark this answers is not in the document]
Comment (jacobwindsor) [reply src/main/java/org/humancellatlas/ingest/submission/SubmissionEnvelope.java:123]: I agree and disagree. Why not do this? Regardless, it was not code that was written in this PR

File: src/main/java/org/humancellatlas/ingest/migrations/MongoChangeLog.java [hunk 6]
```diff
@@ -165,4 +165,22 @@ public void setGraphValidationErrors(MongoDatabase db) {
...
+        filter = Document.parse("{ 'submissionState': 'VALIDATING' }");
+        update = new ArrayList<>();
+        update.add(new Document("$set", Document.parse("{ 'submissionState': 'METADATA_VALIDATING' }")));
+        db.getCollection("submissionEnvelope").updateMany(filter, update);
+
+        filter = Document.parse("{ 'submissionState': 'INVALID' }");
+        update = new ArrayList<>();
+        update.add(new Document("$set", Document.parse("{ 'submissionState': 'METADATA_INVALID' }")));
+        db.getCollection("submissionEnvelope").updateMany(filter, update);
+    }
```
[the remark this answers is not in the document]
Comment (MightyAx) [reply src/main/java/org/humancellatlas/ingest/migrations/MongoChangeLog.java:185]: +1, very neat!

**Review by MightyAx — APPROVED** [review]

**Pull request merged**